pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -0,0 +1,252 @@
+			),
+			readRuleSilenceEvaluator(uid),
+		)
+	}
+
+	// shortcut assertion to check if user can update any silence
+	updateAnySilenceEvaluator = ac.EvalAll(ac.EvalPermission(instancesWrite), readAllSilencesEvaluator)
+	// asserts that user has access to update global silences
+	updateGlobalSilenceEvaluator = ac.EvalAll(ac.EvalPermission(instancesWrite), readSomeSilenceEvaluator)
+	// asserts that user has access to update silences at all
+	updateSomeRuleSilenceEvaluator = ac.EvalAll(
+		readSomeSilenceEvaluator,
+		ac.EvalAny(
+			ac.EvalPermission(instancesWrite),
+			ac.EvalPermission(silenceWrite)),
+	)
+	// asserts that user has access to create silences in a specific folder
+	updateRuleSilenceEvaluator = func(uid string) ac.Evaluator {
+		return ac.EvalAll(
+			ac.EvalAny(
+				ac.EvalPermission(instancesWrite),
+				ac.EvalPermission(silenceWrite, dashboards.ScopeFoldersProvider.GetResourceScopeUID(uid)),
+			),
+			readRuleSilenceEvaluator(uid),
+		)
+	}
+)
+
+type Silence interface {
+	GetRuleUID() *string
+}
+
+type RuleUIDToNamespaceStore interface {
+	GetNamespacesByRuleUID(ctx context.Context, orgID int64, uids ...string) (map[string]string, error)
+}
+
+type SilenceService struct {
+	genericService
+	store RuleUIDToNamespaceStore
+}
+
+func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *SilenceService {
+	return &SilenceService{
+		genericService: genericService{
+			ac: ac,
+		},
+		store: store,
+	}
+}
+
+// FilterByAccess filters the given list of silences based on the access control permissions of the user.
+// Global silence (one that is not attached to a particular rule) is considered available to all users.
+// For silences that are not attached to a rule, are checked against authorization.
+// This method is more preferred when many silences need to be checked.
+func (s SilenceService) FilterByAccess(ctx context.Context, user identity.Requester, silences ...Silence) ([]Silence, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	if err != nil || canAll { // return early if user can either read all silences or there is an error
+		return silences, err
+	}
+	canSome, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
+	if err != nil || !canSome {
+		return nil, err
+	}
+	result := make([]Silence, 0, len(silences))
+	keys := make(map[string][]Silence, len(silences))
+	for _, silence := range silences {
+		ruleUID := silence.GetRuleUID()
+		if ruleUID == nil { // if this is a global silence
+			result = append(result, silence)
+			continue
+		}
+		key := *ruleUID
+		keys[key] = append(keys[key], silence)
+	}
+	if len(keys) == 0 { // if only global silences are provided no need in other checks
+		return result, nil
+	}
+	namespaces, err := s.store.GetNamespacesByRuleUID(ctx, user.GetOrgID(), maps.Keys(keys)...)
+	if err != nil {
+		return nil, err
+	}
+
+	for key, silence := range keys {
+		ns, ok := namespaces[key]
```

Comment:
We're likely hitting the DB more than is necessary here since we don't need to check the same namespace more than once.

Reply:
We will query database once (or more if there are a lot of rule UIDs and we will have to use batch queries, similar to [GetFolders](https://github.com/grafana/grafana/blob/1208888bb618774228aefe65884e06316a06e714/pkg/services/folder/folderimpl/sqlstore.go#L441)), and as the result we get a map of "rule UID to folder UID" which we query here.

Reply:
Ah, I get what you saying. You mean that we check HasAccess multiple times for the same folder. Good catch!